### ISSUE TITLE
2.0b1 Bug: Draftail crash when rich text field is named "description"

### ISSUE THREAD

**Issue by strindhaug** [wagtail/wagtail#4295]
### Issue Summary

If a richtext field is called "description" (and presumably also "viewport") the script fails, because 
the following
```js
const field = document.querySelector(`[name="${fieldName}"]`);
```
matches the first element on the page with the attribute name matching `fieldName`. Since the admin template contains the following
```html
    <meta name="description" content="" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
```
If a rich text field is named "description" or presumably also "viewport" the query selector will match these meta tags instead of the correct input field. Then it will crash when it tries to parse `undefined` at

```js
const rawContentState = JSON.parse(field.value);
```
since the meta tag is not the correct input tag and does not have a value attribute.

### Steps to Reproduce

Make a model with a rich text field named `description`:
```
    description = RichTextField(_('description'))
```
Open the editor

### Technical details
* Python version: 3.6.3 (irrelevant since this is a js bug)
* Django version: 1.11.10 (irrelevant since this is a js bug)
* Wagtail version: Version: 2.0b1
* Browser version: Google Chrome  64.0.3282.140 (though `JSON.parse(undefined)` will fail in any browser)

### Proposed fix

The minimal fix is simply to add the tagname to the querySelector string like so (assuming rich text fields are always input tags):
```js
const field = document.querySelector(`input[name="${fieldName}"]`);
```
(I'm testing this fix now, by modifying the minified file directly on localhost, so I can see if it causes any other issues; if it doesn't I'm happy to make a PR for it later)

Alternatively the script could be modified to use getElementById instead of queryselector, which is slightly faster and less likely to accidentally match a random tag on the page.

**Comment by strindhaug** [comment]
BTW, it would be nice if the script files were build with sourcemaps and the source files available (at least for beta builds like this), so you could jump directly to a sensibly formatted file in the debugger rather than having to grep for it manually. 

Ideally it should have sourcemaps all the time. If the source map file is in a separate file, only referenced in the bottom of the file with a sourcemaps comment, it doesn't really affect the performance at all unless you open the debugger; in which case having a sourcemap file actually improves performance because it's easier on the debugger to jump to source files rather than to a point in a single line, million character minified file.

**Comment by thibaudcolas** [comment]
👍 I'll have a look at this, probably switch to the field's `id` which is more likely to be unique.

I'm not sure why source maps haven't been included so far, it would definitely make debugging much easier. It should be a straightforward change so I'll make a separate PR for that and we can further discuss there.

**Comment by gasman** [comment]
It's important that we don't build front-end assets on beta / RC versions any differently from the final releases, because then we'll miss bugs that occur in the build process, like we did in #3944.

**Comment by strindhaug** [comment]
@gasman  Sure, but I don't see any reason not to have external sourcemaps in production as well. 

And by "external", i mean the kind where only a short comment is added to the minified file like this:
```js
//# sourceMappingURL=somefilename.js.map
```
This adds just a few bytes to the minimized file, but makes debugging an js error a lot easier if it should occur. The only "disadvantage" besides a slightly slower build, is that people can see your unminified source code easily; but this is open source anyway, so...

I think adding `config.devtool = 'source-map';` to the prod file or something should do that; (I have some trouble getting node up and running correctly, so I haven't testet it yet).

**Comment by rspeed** [comment]
Instead of `document.querySelector`, I suggest using `document.forms['form id'].elements['field name']`, since it's an API specific to form elements.

**Comment by rspeed** [comment]
If Wagtail does the leg-work to make sure the element IDs are truly unique, it should be safe to just use `document.getElementById`. Using `document.currentScript` to determine context seems a bit fragile to me, as it would break if the scripts were relocated to somewhere else in the DOM.

**Comment by thibaudcolas** [comment]
I agree that it seems fragile on principle, but this is the one thing Wagtail actually enforces with `WidgetWithScript`.

It would break if this code is refactored willy-nilly, but `document.currentScript` is used at the call site rather than in the init function's implementation so hopefully it's clear enough that script location matters.

As far as I know, there is no particular enforcement of ID uniqueness.

**Comment by rspeed** [comment]
Fair enough! I guess the root of what's bothering me is actually part of Wagtail's design, not this PR.

**Comment by jeromelebleu** [comment]
Hi, I'm far away from Wagtail's source code but @thibaudcolas 's fix works for me. I now have the Draftail editor on my `description` field, thanks!

**Comment by gasman** [comment]
Unfortunately the fix in #4301 has broken the initialisation of newly-added RichTextBlocks in StreamFields (and, I suspect, RichTextFields inside InlinePanels). Will investigate...

**Comment by thibaudcolas** [comment]
Looking into this, it's because jQuery evaluates dynamically inserted scripts [in the `head`](https://github.com/jquery/jquery/blob/625e19cd9be4898939a7c40dbeb2b17e40df9d54/src/core/DOMEval.js#L26), not where the script will actually be inserted, so `document.currentScript` in the editor init gives the wrong context.
